tf_quant_finance/models/heston/approximations/heston.py

```diff
@@ -0,0 +1,326 @@
+    strikes = tf.convert_to_tensor(strikes, dtype=dtype, name='strikes')
+    dtype = strikes.dtype
+    expiries = tf.convert_to_tensor(expiries, dtype=dtype, name='expiries')
+    kappas = tf.convert_to_tensor(kappas, dtype=dtype, name='kappas')
+    thetas = tf.convert_to_tensor(thetas, dtype=dtype, name='thetas')
+    sigmas = tf.convert_to_tensor(sigmas, dtype=dtype, name='sigmas')
+    rhos = tf.convert_to_tensor(rhos, dtype=dtype, name='rhos')
+    variances = tf.convert_to_tensor(variances, dtype=dtype, name='variances')
+    forwards = tf.convert_to_tensor(forwards, dtype=dtype, name='forwards')
+
+    if discount_rates is not None:
+      discount_rates = tf.convert_to_tensor(
+          discount_rates, dtype=dtype, name='discount_rates')
+    elif discount_factors is not None:
+      discount_rates = -tf.math.log(discount_factors) / expiries
+    else:
+      discount_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='discount_rates')
+
+    if continuous_dividends is None:
+      continuous_dividends = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='continuous_dividends')
+
+    if cost_of_carries is not None:
+      cost_of_carries = tf.convert_to_tensor(
+          cost_of_carries, dtype=dtype, name='cost_of_carries')
+    else:
+      cost_of_carries = discount_rates - continuous_dividends
+
+    if discount_factors is not None:
+      discount_factors = tf.convert_to_tensor(
+          discount_factors, dtype=dtype, name='discount_factors')
+    else:
+      discount_factors = tf.exp(-discount_rates * expiries)
+
+    if forwards is not None:
+      forwards = tf.convert_to_tensor(forwards, dtype=dtype, name='forwards')
+    else:
+      spots = tf.convert_to_tensor(spots, dtype=dtype, name='spots')
+      forwards = spots * tf.exp(cost_of_carries * expiries)
+
+    # Cast as complex for the characteristic function calculation
+    expiries_real = tf.complex(expiries, tf.zeros_like(expiries))
+    kappas_real = tf.complex(kappas, tf.zeros_like(kappas))
+    thetas_real = tf.complex(thetas, tf.zeros_like(thetas))
+    sigmas_real = tf.complex(sigmas, tf.zeros_like(sigmas))
+    rhos_real = tf.complex(rhos, tf.zeros_like(rhos))
+    variances_real = tf.complex(variances, tf.zeros_like(variances))
+
+    def char_fun(u):
+      """Using 'second formula' for the (first) characteristic function of
+      log( spot_T / forwards )
+      (noted 'phi_2' in 'The Little Heston Trap', (Albrecher))
+      """
+      u_real = tf.complex(u, tf.zeros_like(u))
+      u_imag = tf.complex(tf.zeros_like(u), u)
+
+      s  = rhos_real * sigmas_real * u_imag
+
+      d = (s - kappas_real) ** 2 - sigmas_real ** 2 * (-u_imag - u_real ** 2)
+      d = tf.math.sqrt(d)
+      g = (kappas_real - s - d) / (kappas_real - s + d)
+      a = kappas_real * thetas_real
+      h = g * tf.math.exp(-d * expiries_real)
+      m = 2 * tf.math.log((1 - h) / (1 - g))
+      C = (a / sigmas_real ** 2) * ((kappas_real - s - d) * expiries_real - m)
```

Comment:
usually we are not allowed to use 1-symbol variables unless they are the same as in the reference paper.  So I trust you this is the case.
Unfortunately, you can't use capital letters for variable names (e.g., "C" or "D"). Please come up with a sensible replacement